pkg/ipamutil/reconciler.go

```diff
@@ -0,0 +1,281 @@
+package ipamutil
+
+import (
+	"context"
+	"fmt"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ipamv1 "sigs.k8s.io/cluster-api/exp/ipam/api/v1beta1"
+	clusterutil "sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/annotations"
+	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/cluster-api/util/predicates"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	ctrlhandler "sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
```

Comment:
the finalizer usage here is a bit confusing to me.

When is ReleaseAddress used? At which time? And Protect?

Usually it helps me to describe on a more verbose way at some comment something like 

"When a new claim arrives, it will do X, then Y, then add a finalizer to guarantee this or that"

But this can be a followup!!!! No need to sweat on it

Reply:
I think there is a lot of disagreement around Finalizer Naming. To me a Finalizer takes care of a specific action/function, and is removed after that task is done. When the same controller handles multiple Finalizers it's technically not necessary to have more than one. I think that's the reason while there is often just some generic finalizer to prevent deletion which gets removed after everything is done.
In this case the Finalizers serve different purposes though. The `ProtectAddress` one prevents deletion until the Claim is deleted. The `ReleaseAddress` one cleans up the allocated address after the claim was deleted. With the in-cluster provider this of course happens immediately, with e.g. Infoblox this can be asynchronous.
Whether this approach makes sense is debatable, especially since the `ProtectAddress` finalizer should also prevent the `ReleaseAddress` finalizer from executing, as we don't want to deallocate the address while it is still in use.

Reply:
ok, makes sense. I was looking into the perspective of incluster and keep forgetting that at some points the deletion may be an async operation